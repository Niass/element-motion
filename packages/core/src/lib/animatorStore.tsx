import { CollectorData, ElementData } from '../Collector';

export interface AnimatorData {
  elementData: ElementData;
  collectorData: CollectorData[];
}

interface AnimatorStore {
  [key: string]: AnimatorData;
}

const store: AnimatorStore = {};

export const set = (key: string, value: AnimatorData) => {
  store[key] = value;
};

export const get = (key: string): AnimatorData => {
  return store[key];
};

export const has = (key: string): boolean => !!get(key);

export const remove = (key: string) => {
  delete store[key];
};
