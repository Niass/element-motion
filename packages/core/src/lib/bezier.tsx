/* eslint-disable no-restricted-properties */
import { Point } from './types';

export const bezier = (start: Point, end: Point, controlPoint1: Point, controlPoint2: Point) => {
  const cX = 3 * (controlPoint1.x - start.x);
  const bX = 3 * (controlPoint2.x - controlPoint1.x) - cX;
  const aX = end.x - start.x - cX - bX;

  const cY = 3 * (controlPoint1.y - start.y);
  const bY = 3 * (controlPoint2.y - controlPoint1.y) - cY;
  const aY = end.y - start.y - cY - bY;

  return (time: number): Point => {
    const x = aX * Math.pow(time, 3) + bX * Math.pow(time, 2) + cX * time + start.x;
    const y = aY * Math.pow(time, 3) + bY * Math.pow(time, 2) + cY * time + start.y;

    return { x, y };
  };
};
