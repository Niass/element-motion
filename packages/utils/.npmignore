examples
test
tsconfig*
*test*
*stories*
__docz__
__docs__
src
dist/**/package.json
