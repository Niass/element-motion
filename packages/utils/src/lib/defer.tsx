const defer = <TValue extends void>() => {
  let resolve: (value?: TValue | PromiseLike<TValue> | undefined) => void = () => {};
  let reject: (value?: TValue | PromiseLike<TValue> | undefined) => void = () => {};

  const promise = new Promise<TValue>((res, rej) => {
    resolve = res;
    reject = rej;
  });

  return {
    resolve,
    reject,
    promise,
  };
};

export default defer;
