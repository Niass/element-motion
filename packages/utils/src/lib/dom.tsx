export function eventListener<K extends keyof HTMLElementEventMap>(
  element: HTMLElement,
  event: K,
  cb: (this: HTMLElement, ev: HTMLElementEventMap[K]) => any,
  options?: boolean | AddEventListenerOptions | undefined
) {
  element.addEventListener<K>(event, cb, options);
  return () => element.removeEventListener(event, cb, options);
}

export function getDocumentScroll() {
  const scrollTop =
    document.documentElement && document.documentElement.scrollTop
      ? document.documentElement.scrollTop
      : document.body && document.body.scrollTop;

  const scrollLeft =
    document.documentElement && document.documentElement.scrollLeft
      ? document.documentElement.scrollLeft
      : document.body && document.body.scrollLeft;

  return {
    scrollTop: scrollTop || 0,
    scrollLeft: scrollLeft || 0,
  };
}

export interface ElementBoundingBoxOpts {
  useOffsetSize?: boolean;
}

export interface ElementBoundingBox {
  size: {
    width: number;
    height: number;
  };
  location: {
    left: number;
    top: number;
  };
  raw: {
    rect: ClientRect | DOMRect;
    scrollTop: number;
    scrollLeft: number;
  };
}

export function getElementBoundingBox(
  element: HTMLElement,
  options: ElementBoundingBoxOpts = {}
): ElementBoundingBox {
  const rect = element.getBoundingClientRect();
  const { scrollLeft, scrollTop } = getDocumentScroll();

  return {
    size: {
      width: options.useOffsetSize ? element.offsetWidth : rect.width,
      height: options.useOffsetSize ? element.offsetHeight : rect.height,
    },
    location: {
      left: rect.left + scrollLeft,
      top: rect.top + scrollTop,
    },
    raw: {
      rect,
      scrollTop,
      scrollLeft,
    },
  };
}

export function calculateElementCenterInViewport(elementBoundingBox: ElementBoundingBox) {
  return {
    top: elementBoundingBox.location.top + Math.ceil(elementBoundingBox.size.width / 2),
    left: elementBoundingBox.location.left - Math.ceil(elementBoundingBox.size.height / 2),
  };
}

export function getWindowDimensions() {
  return {
    width: window.innerWidth,
    height: window.innerHeight,
  };
}

export function calculateWindowCentre() {
  return {
    left: Math.ceil(window.innerWidth / 2),
    top: Math.ceil(window.innerHeight / 2),
  };
}

export function recalculateElementBoundingBoxFromScroll(
  elementBoundingBox: ElementBoundingBox
): ElementBoundingBox {
  const { scrollTop, scrollLeft } = getDocumentScroll();
  const scrollTopDiff = scrollTop - elementBoundingBox.raw.scrollTop;
  const scrollLeftDiff = scrollLeft - elementBoundingBox.raw.scrollLeft;

  return {
    ...elementBoundingBox,
    location: {
      top: elementBoundingBox.location.top + scrollTopDiff,
      left: elementBoundingBox.location.left + scrollLeftDiff,
    },
  };
}
