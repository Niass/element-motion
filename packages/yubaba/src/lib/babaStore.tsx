import { CollectorData, ElementData } from '../Collector';

export interface BabaData {
  elementData: ElementData;
  collectorData: CollectorData[];
}

interface BabaStore {
  [key: string]: BabaData;
}

const store: BabaStore = {};

export const set = (key: string, value: BabaData) => {
  store[key] = value;
};

export const get = (key: string): BabaData => {
  return store[key];
};

export const has = (key: string): boolean => !!get(key);

export const remove = (key: string) => {
  delete store[key];
};
