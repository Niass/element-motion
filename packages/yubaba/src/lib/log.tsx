import pkg from '../../package.json';

export const precondition = (check: any, message: string) => {
  if (!check) {
    throw new Error(`yubaba v${pkg.version}

${message}`);
  }
};

export const warn = (message: string) => {
  console.warn(`yubaba v${pkg.version}

${message}`);
};
