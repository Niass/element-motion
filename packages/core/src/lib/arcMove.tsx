/*
 Copyright 2016-present The Material Motion Authors. All Rights Reserved.

 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at

 http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
 */

import { ElementBoundingBox } from './dom';
import { Point } from './types';
import { bezier } from './bezier';

interface ControlPoints {
  point1: Point;
  point2: Point;
}

const zeroPoint = { x: 0, y: 0 };
const defaultMinArcAngle: number = 10.0;
const defaultMaxArcAngle: number = 90.0;

function rad2deg(radians: number): number {
  return (radians * 180.0) / Math.PI;
}

function deg2rad(degrees: number): number {
  return (degrees * Math.PI) / 180.0;
}

function distance(from: Point, to: Point): number {
  const deltaX = to.x - from.x;
  const deltaY = to.y - from.y;
  return Math.sqrt(deltaX * deltaX + deltaY * deltaY);
}

function normalized(point: Point): Point {
  const length = Math.sqrt(point.x * point.x + point.y * point.y);
  if (length < Number.EPSILON) {
    return zeroPoint;
  }
  return { x: point.x / length, y: point.y / length };
}

export function getControlPoints(from: Point, to: Point): ControlPoints {
  if (from.x === to.x && from.y === to.y) {
    return { point1: from, point2: to };
  }

  const minArcAngleRad = deg2rad(defaultMinArcAngle);

  const deltaX = to.x - from.x;
  const deltaY = to.y - from.y;

  const pointC = { x: to.x, y: from.y };
  const pointD = { x: (from.x + to.x) * 0.5, y: (from.y + to.y) * 0.5 };

  // Calculate side lengths
  const lenAB = distance(from, to);
  const lenAC = distance(from, pointC);
  const lenBC = distance(to, pointC);

  // Length AD is half length AB
  const lenAD = lenAB * 0.5;

  // Angle alpha
  let alpha: number;
  if (Math.abs(deltaX) > Math.abs(deltaY)) {
    alpha = Math.acos(lenAC / lenAB);
  } else {
    alpha = Math.acos(lenBC / lenAB);
  }

  // Alpha in degrees
  const alphaDeg = rad2deg(alpha);

  // Calculate point E
  const lenAE = lenAD / Math.cos(alpha);
  let pointE: Point;
  if (from.y === to.y) {
    pointE = pointD;
  } else if (Math.abs(deltaX) > Math.abs(deltaY)) {
    const normalizedCFrom = normalized({ x: pointC.x - from.x, y: pointC.y - from.y });
    pointE = { x: from.x + normalizedCFrom.x * lenAE, y: from.y + normalizedCFrom.y * lenAE };
  } else {
    const normalizedCTo = normalized({ x: pointC.x - to.x, y: pointC.y - to.y });
    pointE = { x: to.x + normalizedCTo.x * lenAE, y: to.y + normalizedCTo.y * lenAE };
  }

  // Constrain DE to account for min/max arc segment
  const arcAngleClampDeg = Math.min(
    defaultMaxArcAngle,
    Math.max(defaultMinArcAngle, alphaDeg * 2.0)
  );

  const arcAngleClamp = deg2rad(arcAngleClampDeg);
  const alphaClamp = arcAngleClamp / 2.0;
  const maxLen = lenAD * Math.tan(alphaClamp);

  // Point E'
  let pointE2: Point;
  const vDE: Point = { x: pointE.x - pointD.x, y: pointE.y - pointD.y };
  const lenDE = distance(zeroPoint, vDE);

  let adjMinLen: number;
  if (defaultMinArcAngle > 0) {
    const tanMinArcAngleRad = Math.tan(minArcAngleRad);
    if (Math.abs(tanMinArcAngleRad) < Number.EPSILON) {
      // Protection against possible divide by zero - shouldn't happen in practice.
      adjMinLen = Number.MAX_SAFE_INTEGER;
    } else {
      const lenADOverTanMinArcAngleRad = lenAD / tanMinArcAngleRad;
      adjMinLen =
        Math.sqrt(lenAD * lenAD + lenADOverTanMinArcAngleRad ** 2) - lenADOverTanMinArcAngleRad;
    }
  } else {
    adjMinLen = 0;
  }
  if (Math.abs(deltaY) > Math.abs(deltaX)) {
    adjMinLen = Math.max(0, Math.min(lenDE, maxLen));
  }

  const newLen = Math.max(adjMinLen, Math.min(lenDE, maxLen));
  if (from.y === to.y) {
    pointE2 = { x: pointD.x, y: pointD.y + newLen };
  } else {
    const normalizedVDE = normalized(vDE);
    pointE2 = { x: pointD.x + normalizedVDE.x * newLen, y: pointD.y + normalizedVDE.y * newLen };
  }

  // Alpha'
  const lenDE2 = distance(pointD, pointE2);
  const alpha2 = Math.atan(lenDE2 / lenAD);

  // Alpha' degrees.
  const alpha2deg = rad2deg(alpha2);

  // Beta' degrees.
  const beta2deg = 90.0 - alpha2deg;

  // Beta'.
  const beta2 = deg2rad(beta2deg);

  // Radius'.
  const radius2 = lenAD / Math.cos(beta2);

  // Calculate the cubic bezier tangent handle length
  //
  // The following method is for a 90 degree arc
  //
  // tangent length = radius * k * scaleFactor
  //
  // radius: radius of our circle
  // kappa: constant with value of ~0.5522847498
  // scaleFactor: proportion of our arc to a 90 degree arc (arc angle / 90)
  const kappa: number = 0.5522847498;
  const radScaling: number = (alpha2deg * 2.0) / 90.0;
  const tangentLength = radius2 * kappa * radScaling;

  // Calculate the in tangent position in world coordinates
  // The tangent handle lies along the line between points B and E'
  // with magnitude of tangentLength
  const vBEnorm = normalized({ x: pointE2.x - to.x, y: pointE2.y - to.y });
  const inTangent: Point = {
    x: to.x + vBEnorm.x * tangentLength,
    y: to.y + vBEnorm.y * tangentLength,
  };

  // Calculate the out tangent position in world coordinates
  // The tangent handle lies along the line between points A and E'
  // with magnitude of tangentLength
  const vAEnorm = normalized({ x: pointE2.x - from.x, y: pointE2.y - from.y });
  const outTangent = {
    x: from.x + vAEnorm.x * tangentLength,
    y: from.y + vAEnorm.y * tangentLength,
  };

  return { point1: outTangent, point2: inTangent };
}

export function arcMove(origin: ElementBoundingBox, destination: ElementBoundingBox) {
  const startPoint = { x: origin.location.left, y: origin.location.top };
  const endPoint = { x: destination.location.left, y: destination.location.top };
  const controlPoints = getControlPoints(startPoint, endPoint);
  const getBezierPoint = bezier(startPoint, endPoint, controlPoints.point1, controlPoints.point2);
  const pointsAccuracy = 1 / 100;
  const points = [];

  for (let time = 0; time < 1; time += pointsAccuracy) {
    points.push(getBezierPoint(time));
  }

  return points;
}
