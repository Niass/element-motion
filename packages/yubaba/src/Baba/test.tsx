import * as React from 'react';
import { mount } from 'enzyme';
import 'jest-enzyme';
import { WrappedBaba as Baba } from '../Baba';
import Target from '../FocalTarget';
import { getElementBoundingBox } from '../lib/dom';
import defer from '../lib/defer';
import * as utils from '../__tests__/utils';

jest.mock('../lib/dom');

describe('<Baba />', () => {
  it('should callback when animation has finished', done => {
    const Animation = utils.createTestAnimation();
    const wrapper = mount(
      <utils.BabaUnderTest
        from={
          <Baba name="anim-0">
            <Animation>{props => <div {...props} />}</Animation>
          </Baba>
        }
        to={
          <Baba name="anim-0" onFinish={done}>
            <div />
          </Baba>
        }
        start={false}
      />
    );

    wrapper.setProps({
      start: true,
    });
  });

  it('should pass dom data to child animation', async () => {
    const callback = jest.fn();
    const elementData = utils.domData();
    (getElementBoundingBox as jest.Mock).mockReturnValue(elementData);
    const deferred = defer();
    const Animation = utils.createTestAnimation({
      onAnimate: callback,
    });
    const wrapper = mount(
      <utils.BabaUnderTest
        from={
          <Baba name="anim-1">
            <Animation>{props => <div {...props} />}</Animation>
          </Baba>
        }
        to={
          <Baba name="anim-1" onFinish={deferred.resolve}>
            <div />
          </Baba>
        }
        start={false}
      />
    );

    wrapper.setProps({
      start: true,
    });
    await deferred.promise;

    expect(callback.mock.calls[0]).toMatchSnapshot();
  });

  it('should pass dom data to child animation when using in prop', async () => {
    const callback = jest.fn();
    const elementData = utils.domData();
    (getElementBoundingBox as jest.Mock).mockReturnValue(elementData);
    const deferred = defer();
    const Animation = utils.createTestAnimation({
      onAnimate: callback,
    });
    const wrapper = mount(
      <utils.BabaUnderTest
        from={start => (
          <Baba name="anim-aa" in={!start}>
            <Animation>{props => <div {...props} />}</Animation>
          </Baba>
        )}
        to={
          <Baba name="anim-aa" onFinish={deferred.resolve}>
            <div />
          </Baba>
        }
        start={false}
      />
    );

    wrapper.setProps({
      start: true,
    });
    await deferred.promise;

    expect(callback.mock.calls[0]).toMatchSnapshot();
  });

  it('should pass target dom data to child animation', async () => {
    const callback = jest.fn();
    const elementData = utils.domData();
    (getElementBoundingBox as jest.Mock).mockReturnValue(elementData);
    const deferred = defer();
    const Animation = utils.createTestAnimation({
      onAnimate: callback,
    });
    const wrapper = mount(
      <utils.BabaUnderTest
        from={
          <Baba name="anim-bb">
            <Animation>{props => <div {...props} />}</Animation>
          </Baba>
        }
        to={
          <Baba name="anim-bb" onFinish={deferred.resolve}>
            <Animation>
              {animProps => (
                <Target>
                  {props => (
                    <main {...animProps}>
                      <div {...props} />
                    </main>
                  )}
                </Target>
              )}
            </Animation>
          </Baba>
        }
        start={false}
      />
    );

    wrapper.setProps({
      start: true,
    });
    await deferred.promise;

    expect(callback.mock.calls[0][0].destination.elementBoundingBox).toMatchSnapshot();
  });
});
