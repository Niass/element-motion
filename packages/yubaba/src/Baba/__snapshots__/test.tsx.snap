// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Baba /> should pass dom data to child animation 1`] = `
Array [
  Object {
    "destination": Object {
      "element": undefined,
      "elementBoundingBox": Object {
        "location": Object {
          "left": 0,
          "top": 0,
        },
        "raw": Object {
          "rect": Object {},
          "scrollLeft": 0,
          "scrollTop": 0,
        },
        "size": Object {
          "height": 200,
          "width": 100,
        },
      },
      "focalTargetElement": undefined,
      "focalTargetElementBoundingBox": undefined,
      "render": undefined,
    },
    "origin": Object {
      "element": <div />,
      "elementBoundingBox": Object {
        "location": Object {
          "left": 0,
          "top": 0,
        },
        "raw": Object {
          "rect": Object {},
          "scrollLeft": 0,
          "scrollTop": 0,
        },
        "size": Object {
          "height": 200,
          "width": 100,
        },
      },
      "focalTargetElement": undefined,
      "focalTargetElementBoundingBox": undefined,
      "render": [Function],
    },
  },
]
`;

exports[`<Baba /> should pass dom data to child animation when using in prop 1`] = `
Array [
  Object {
    "destination": Object {
      "element": undefined,
      "elementBoundingBox": Object {
        "location": Object {
          "left": 0,
          "top": 0,
        },
        "raw": Object {
          "rect": Object {},
          "scrollLeft": 0,
          "scrollTop": 0,
        },
        "size": Object {
          "height": 200,
          "width": 100,
        },
      },
      "focalTargetElement": undefined,
      "focalTargetElementBoundingBox": undefined,
      "render": undefined,
    },
    "origin": Object {
      "element": <div />,
      "elementBoundingBox": Object {
        "location": Object {
          "left": 0,
          "top": 0,
        },
        "raw": Object {
          "rect": Object {},
          "scrollLeft": 0,
          "scrollTop": 0,
        },
        "size": Object {
          "height": 200,
          "width": 100,
        },
      },
      "focalTargetElement": undefined,
      "focalTargetElementBoundingBox": undefined,
      "render": [Function],
    },
  },
]
`;

exports[`<Baba /> should pass target dom data to child animation 1`] = `
Object {
  "location": Object {
    "left": 0,
    "top": 0,
  },
  "raw": Object {
    "rect": Object {},
    "scrollLeft": 0,
    "scrollTop": 0,
  },
  "size": Object {
    "height": 200,
    "width": 100,
  },
}
`;
